### PR TITLE
Restore TF embeddings and attention layers to their previous version

### PR DESCRIPTION
**Pull request by jplu**
# What does this PR do?

This PR restores the attention layers and the embeddings as it was in v4.1, even though the embeddings got few improvements compared to their original version to keep XLA compliancy. The reason is because we realized that some used operators were not compatible with some NN SDKs such as the one from Qualcomm or ONNX.

### CODE REVIEW

File: src/transformers/models/bert/modeling_tf_bert.py [hunk 1]
```diff
@@ -314,31 +213,26 @@ def __init__(self, config: BertConfig, **kwargs):
...
-            bias_axes="de",
-            kernel_initializer=get_initializer(config.initializer_range),
-            name="value",
+        self.value = tf.keras.layers.Dense(
+            units=self.all_head_size, kernel_initializer=get_initializer(config.initializer_range), name="value"
         )
         self.dropout = tf.keras.layers.Dropout(rate=config.attention_probs_dropout_prob)
 
+    def transpose_for_scores(self, tensor: tf.Tensor, batch_size: int) -> tf.Tensor:
+        tensor = tf.reshape(tensor=tensor, shape=(batch_size, -1, self.num_attention_heads, self.attention_head_size))
```
Comment (sgugger) [inline src/transformers/models/bert/modeling_tf_bert.py:232]: A comment here explaining this goes from `batch size x seq_length x (num_head * head_size)` to `batch_size x num_head x seq_length x head_size` would be useful IMO.

**Review by sgugger — APPROVED** [review]
Thanks a lot for doing this! Just have one small comment.

**Comment by LysandreJik** [comment]
Pinging @mfuntowicz

File: src/transformers/models/bert/modeling_tf_bert.py [hunk 2]
```diff
@@ -347,15 +241,21 @@ def call(
...
 
         # Take the dot product between "query" and "key" to get the raw
         # attention scores.
+        # (batch size, num_heads, seq_len_q, seq_len_k)
+        attention_scores = tf.matmul(query_layer, key_layer, transpose_b=True)
         dk = tf.cast(self.attention_head_size, dtype=query_layer.dtype)
-        query_layer = tf.multiply(query_layer, tf.math.rsqrt(dk))
-        attention_scores = tf.einsum("aecd,abcd->acbe", key_layer, query_layer)
+        one_cst = tf.constant(1.0, dtype=dk.dtype)
+        query_layer = tf.multiply(query_layer, tf.divide(one_cst, tf.math.sqrt(dk)))
```
Comment (mfuntowicz) [inline src/transformers/models/bert/modeling_tf_bert.py:258]: Cache this `1.0 / sqrt(...)` in the initialization method or build? 

This doesn't change in between calls, so you save some time not re-computing.

Comment (jplu) [reply src/transformers/models/bert/modeling_tf_bert.py:258]: This indeed can be put in the init, I will update this 👍  But cannot be done in plain Python, because the dtype has to be compliant in case of mixed precision defined to something else than `float32`.

File: src/transformers/models/bert/modeling_tf_bert.py [hunk 3]
```diff
@@ -372,7 +272,11 @@ def call(
         if head_mask is not None:
             attention_probs = tf.multiply(attention_probs, head_mask)
```
Comment (mfuntowicz) [inline src/transformers/models/bert/modeling_tf_bert.py:273]: In the continuity of @sgugger comments on `transpose_for_score` do we really need to split the dimension if we are not going to mask some heads? 

I wonder if the calls to `transpose_for_scores` cannot be moved only in this if branch to avoid unecessary reshaping operations on the common path.

File: src/transformers/models/bert/modeling_tf_bert.py (hunk 2 above)
Comment (mfuntowicz) [reply src/transformers/models/bert/modeling_tf_bert.py:258]: Can't we compute over plain old python and store the results _(in `init`)_ with `tf.constant(..., dtype=...)`

File: src/transformers/models/bert/modeling_tf_bert.py (hunk 3 above)
Comment (jplu) [reply src/transformers/models/bert/modeling_tf_bert.py:273]: All the reshape can be even entirely removed if we introduce our own way to compute a 3D dense. This should not be too difficult.

File: src/transformers/models/bert/modeling_tf_bert.py (hunk 2 above)
Comment (jplu) [reply src/transformers/models/bert/modeling_tf_bert.py:258]: We can absolutely doing that 👍 We can detect any policy update with `self._compute_dtype` in a layer.

Comment (jplu) [reply src/transformers/models/bert/modeling_tf_bert.py:258]: Ah finally no, I found a use case for which it fails. If you run `tf.keras.mixed_precision.experimental.set_policy('mixed_float16')` before to instantiate your model, it works, if you do it after it fails. So it is safer to have the cast inside the `call`.

File: src/transformers/models/bert/modeling_tf_bert.py [hunk 4]
```diff
@@ -314,31 +214,28 @@ def __init__(self, config: BertConfig, **kwargs):
                 f"of attention heads ({config.num_attention_heads})"
             )
 
+        self.num_attention_heads = config.num_attention_heads
         self.attention_head_size = int(config.hidden_size / config.num_attention_heads)
+        self.all_head_size = self.num_attention_heads * self.attention_head_size
+        self.rsqrt_att_head_size = tf.cast(1.0 / math.sqrt(self.attention_head_size), dtype=self._compute_dtype)
```
Comment (sgugger) [inline src/transformers/models/bert/modeling_tf_bert.py:220]: No I don't think this will work -> the dtype could change if the user changes their policy after instantiating the model. We should remove the `dtype = self._compute_dtype` and do the `cast` in the call as was done before.
(Relying on the [mixed precision doc](https://www.tensorflow.org/api_docs/python/tf/keras/mixed_precision/Policy#how_to_write_a_layer_that_supports_mixed_precision_and_float64_2) here.)

Comment (jplu) [reply src/transformers/models/bert/modeling_tf_bert.py:220]: This is what I answered in the Morgan's comment https://github.com/huggingface/transformers/pull/9890#discussion_r567723981

Comment (sgugger) [reply src/transformers/models/bert/modeling_tf_bert.py:220]: Yes, and I disagree with it. This is not how the documentation indicate we should treat new tensors values and we should not assume the user is always setting the right policy before instantiating their model.

File: src/transformers/models/bert/modeling_tf_bert.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/models/bert/modeling_tf_bert.py:235]: I meant for the whole function, not just this reshape ;-)

File: src/transformers/models/bert/modeling_tf_bert.py [hunk 5]
```diff
@@ -314,31 +214,28 @@ def __init__(self, config: BertConfig, **kwargs):
                 f"of attention heads ({config.num_attention_heads})"
             )
 
+        self.num_attention_heads = config.num_attention_heads
         self.attention_head_size = int(config.hidden_size / config.num_attention_heads)
+        self.all_head_size = self.num_attention_heads * self.attention_head_size
+        self.rsqrt_att_head_size = 1.0 / math.sqrt(self.attention_head_size)
```
Comment (sgugger) [inline src/transformers/models/bert/modeling_tf_bert.py:220]: Not sure if @mfuntowicz shared it with you, but he observed some instability when using this version, whereas sorting the `sqrt` (just `math.sqrt(self.attention_head_size)`) and then using `tf.divide` in the call instead of `tf.multiply` was more stable.

Comment (jplu) [reply src/transformers/models/bert/modeling_tf_bert.py:220]: Ok, let's put this instead then :)

Comment (jplu) [reply src/transformers/models/bert/modeling_tf_bert.py]: Ok, now it should be ok!

**Review by LysandreJik — APPROVED** [review]
The changes look good to me, thank you!

Morgan was mentioning that the `transpose_for_score` method was called right after the Q/K/V projection, but that there was no need to split this dimension if we're not doing head masking. 
What do you think? Maybe that's some work for another PR, though.

**Comment by jplu** [comment]
> Morgan was mentioning that the transpose_for_score method was called right after the Q/K/V projection, but that there was no need to split this dimension if we're not doing head masking.
What do you think? Maybe that's some work for another PR, though.

I think it seems doable but not sure, I prefer to keep things like this to be sure we revert properly as it was before and we get at least a proper version, and we can take care of this change in another PR.

**Comment by LysandreJik** [comment]
Sounds good to me!

**Comment by jplu** [comment]
LGTM! @patrickvonplaten feel free to merge if you approve the changes ^^

**Comment by patrickvonplaten** [comment]
I won't have time to do a proper review today (can do it tomorrow), but feel free to merge without me if @LysandreJik and @sgugger are ok with it

**Comment by LysandreJik** [comment]
@patrickvonplaten if you can take a look at it today and merge it if it's fine with you, that would be great

File: src/transformers/models/bert/modeling_tf_bert.py [hunk 6]
```diff
@@ -285,18 +184,19 @@ def call(
...
 
         if token_type_ids is None:
-            input_shape = shape_list(inputs_embeds)[:-1]
             token_type_ids = tf.fill(dims=input_shape, value=0)
 
         if position_ids is None:
-            position_embeds = self.position_embeddings(inputs_embeds)
-        else:
-            position_embeds = self.position_embeddings(position_ids)
+            position_ids = tf.range(start=0, limit=input_shape[-1])[tf.newaxis, :]
```
Comment (mfuntowicz) [inline src/transformers/models/bert/modeling_tf_bert.py:195]: Can we use `tf.expand_dims(..., axis=0)` rather than using the strided slice notation? It's a numpy compatibility and not very well supported by potential ISV

Comment (jplu) [reply src/transformers/models/bert/modeling_tf_bert.py:195]: I was recently thinking the same thing, "great minds think alike." 😄
